modules/onewire/onewire.c

```diff
@@ -0,0 +1,439 @@
+    return ONEWIRE_SUCCESS;
+}
+
+OneWireStatus OneWire_Output(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return ONEWIRE_FAILURE;
+    palSetLineMode(OneWireStruct->PalLine, PAL_STM32_MODE_OUTPUT | PAL_STM32_OSPEED_HIGHEST);
+    return ONEWIRE_SUCCESS;
+}
+
+// ONEWIRE IMPLEMENTATION
+
+OneWireStatus OneWire_Init(OneWire_t* OneWireStruct, uint32_t PalLine) {
+    if (!OneWireStruct) return ONEWIRE_FAILURE;
+    OneWireStruct->PalLine = PalLine;
+    return ONEWIRE_SUCCESS;
+}
+
+OneWireStatus OneWire_Reset(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return ONEWIRE_FAILURE;
+    uint8_t presence_pulse = 1;
+    /* Line low, and wait */
+    OneWire_Output(OneWireStruct);
+    OneWire_Low(OneWireStruct);
+    OneWire_Delay(ONEWIRE_TX_MIN_RESET_PULSE_TIME_USEC);
+    
+    /* Release line and wait */
+    OneWire_Input(OneWireStruct);
+    OneWire_Delay(ONEWIRE_WAIT_PRESENCE_PULSE_TIME_USEC);
+    
+    /* Check bit value */
+    presence_pulse = palReadLine(OneWireStruct->PalLine);
+
+    /* Complete reset pulse min time */
+    OneWire_Delay(ONEWIRE_RX_MIN_RESET_PULSE_TIME_USEC);
+    /* Return value of presence pulse, 0 = OK, 1 = ERROR */
+    if (presence_pulse == 0){
+        return ONEWIRE_SUCCESS;
+    }
+    return ONEWIRE_FAILURE;
+}
+
+OneWireStatus OneWire_WriteBit(OneWire_t* OneWireStruct, uint8_t bit) {
+    if (!OneWireStruct) return ONEWIRE_FAILURE;
+    if (bit) {
+        /* Set line low */
+        OneWire_Output(OneWireStruct);
+        OneWire_Low(OneWireStruct);
+        OneWire_Delay(ONEWIRE_TX_WRITE_1_BIT_LO_TIME_USEC);
+
+        /* release line, Hold for remainder of min time for write 1 slot */
+        OneWire_Input(OneWireStruct);
+        OneWire_Delay(ONEWIRE_TX_WRITE_1_BIT_HI_TIME_USEC);
+    } else {
+        /* Set line low */
+        OneWire_Output(OneWireStruct);
+        OneWire_Low(OneWireStruct);
+        OneWire_Delay(ONEWIRE_TX_WRITE_0_BIT_LO_TIME_USEC);  
+
+        /* release line, wait for nonzero recovery time */
+        OneWire_Input(OneWireStruct);
+        OneWire_Delay(ONEWIRE_TX_RECOVER_TIME_USEC);
+    }
+    return ONEWIRE_SUCCESS;
+}
+
+OneWireStatus OneWire_ReadBit(OneWire_t* OneWireStruct, uint8_t* ReadVal) {
+    if (!OneWireStruct || !ReadVal) return ONEWIRE_FAILURE;
+    uint8_t bit = 0;
+    /* Line low */
+    OneWire_Output(OneWireStruct);
+    OneWire_Low(OneWireStruct);
+    OneWire_Delay(ONEWIRE_RX_READ_BIT_LO_TIME_USEC); 
+
+    /* Release line */
+    OneWire_Input(OneWireStruct);
+    OneWire_Delay(ONEWIRE_RX_READ_BIT_WAIT_BEFORE_SAMPLE_TIME_USEC);
+
+    /* Read line value near end of 15us */
+    if (palReadLine(OneWireStruct->PalLine)) {
+        /* Bit is HIGH */
+        bit = 1;
+    }
+
+    /* Wait 50us to complete minimum 60us period */
+    OneWire_Delay(50);
+
+    *ReadVal = bit;
+    return ONEWIRE_SUCCESS;
+}
+
+OneWireStatus OneWire_WriteByte(OneWire_t* OneWireStruct, uint8_t byte) {

```

Comment:
It may be a good idea to benchmark this operation.

Reply:
I will take some measurements today.